subprojects/core/src/integTest/groovy/org/gradle/caching/internal/packaging/BuildCacheEntryPackingIntegrationTest.groovy

```diff
@@ -0,0 +1,115 @@
+import spock.lang.Issue
+import spock.lang.Unroll
+
+class BuildCacheEntryPackingIntegrationTest extends DaemonIntegrationSpec implements DirectoryBuildCacheFixture {
+
+    private static final NON_ASCII_NAME = [
+        "ascii-only": "ascii",
+        "space": " ",
+        "zwnj": "\u200c",
+        "chinese": "敏捷的棕色狐狸跳过了懒狗",
+        "cyrillic": "здравствуйте",
+        "hungarian": "Árvíztűrő tükörfúrógép",
+    ].values().join("-")
+
+    private static final DEFAULT_ENCODINGS = [
+        "UTF-8",
+        "ISO-8859-1",
+        "windows-1250",
+    ]
+
+    @Unroll
+    def "can store and load files having non-ascii characters in file name using default file encoding #fileEncoding"() {
+        def fileName = NON_ASCII_NAME + ".txt"
+        def outputFile = file("dir", fileName)
+
+        buildFile << """
+            println "> Default charset: \${java.nio.charset.Charset.defaultCharset()}"
+            println "> Storing file in cache: $fileName"
+
+            task createFile {
+                outputs.dir("dir")
+                outputs.cacheIf { true }
+                doLast {
+                    file("dir/$fileName").text = "output"
+                }
+            }
+        """
+
+        when:
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        executedAndNotSkipped(":createFile")
+
+        when:
+        assert outputFile.delete()
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        skipped(":createFile")
+
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        outputFile.text == "output"
+
+        where:
+        fileEncoding << DEFAULT_ENCODINGS
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9877")
+    @Unroll
+    def "can store and load files having non-ascii characters in property name using default file encoding #fileEncoding"() {
+        def outputFile = file("output.txt")
+
+        buildFile << """
+            println "> Default charset: \${java.nio.charset.Charset.defaultCharset()}"
+            println "> Storing with property name: $NON_ASCII_NAME"
+
+            task createFile {
+                outputs.file("output.txt")
+                    .withPropertyName("$NON_ASCII_NAME")
+                outputs.cacheIf { true }
+                doLast {
+                    file("output.txt").text = "output"
+                }
+            }
+        """
+
+        when:
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        executedAndNotSkipped(":createFile")
+
+        when:
+        assert outputFile.delete()
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        skipped(":createFile")
+
+        then:
+        output.contains("> Default charset: $fileEncoding")
```

Comment:
Consider using `outputContains` instead to have a more compact error in case of failure, since having `output` here will list the content of the file on failure (which is already listed anyway)

Reply:
Hm, I never noticed that method provides better output. I’ll check it out, thanks!

Reply:
I tested it, and `output.contains()` prints a lot less noise than `outputContains()` (47 vs 150 lines). :o

Reply:
Both seem very verbose to me, just wanted to point out that there's a difference.
Both will print the output several times, the difference is in the error message:

<img width="332" alt="Screen Shot 2019-07-16 at 11 47 12" src="https://user-images.githubusercontent.com/1841944/61284585-8ac0e400-a7bf-11e9-9370-03ff2b8b403d.png">

vs

<img width="339" alt="Screen Shot 2019-07-16 at 11 44 31" src="https://user-images.githubusercontent.com/1841944/61284386-20a83f00-a7bf-11e9-9484-352db38e466f.png">

---

On an unrelated topic, maybe we could compact the output of `output.contains` to make it less noisy.

Reply:
We could. Please open a separate issue if you think we should.

Reply:
Opened https://github.com/gradle/gradle/issues/9966#issue-468564257